File: test/Server/test_EncryptedSocket.py
"""Integration tests for EncryptedSocket.

The Python client implements the same wire protocol as the C++ EncryptedSocket class:
- AES-128-GCM with a 96-bit (12-byte) authentication tag
- 8-byte (64-bit) little-endian nonce counter
- Message format: [marker:4][length:4][ciphertext][tag:12]
- Client-to-server nonce starts at 0
- Server-to-client nonce starts at 0x8000000000000000
"""
import socket
import struct
from pathlib import Path

import pytest

from conftest import (
    connect_with_retry,
    require_supported,
    run_scenario,
    reserve_tcp_port,
)

try:
    from cryptography.hazmat.primitives.ciphers import Cipher, algorithms, modes
    _CRYPTO_AVAILABLE = True
except ImportError:
    _CRYPTO_AVAILABLE = False


pytestmark = pytest.mark.skipif(
    not _CRYPTO_AVAILABLE,
    reason="cryptography package is required for EncryptedSocket tests",
)

# ─── Protocol constants ───────────────────────────────────────────────────────

INCOMING_MARKER = 0x50dd3ac6  # client → server begin marker
OUTGOING_MARKER = 0xe859c294  # server → client begin marker
HEADER_SIZE = 8                # marker(4) + length(4)
TAG_SIZE = 12

ZERO_KEY = bytes(16)                  # initial key: all zeros
TEST_KEY = bytes(range(1, 17))        # non-zero test key: 0x01..0x10


# ─── EncryptedClient helper ───────────────────────────────────────────────────

class EncryptedClient:
    """Python-side implementation of the EncryptedSocket wire protocol."""

    def __init__(self, sock: socket.socket, key: bytes = ZERO_KEY) -> None:
        self.sock = sock
        self.sock.settimeout(3.0)
        self.key = key
        self.send_nonce: int = 0
        self.recv_nonce: int = 0x8000000000000000

    def set_key(self, key: bytes) -> None:
        self.key = key

    def send_message(self, plaintext: bytes) -> None:
        """Encrypt and send one message to the server."""
        nonce = struct.pack("<Q", self.send_nonce)
        self.send_nonce += 1
        msg_length = HEADER_SIZE + len(plaintext) + TAG_SIZE
        aad = struct.pack("<I", msg_length)

        enc = Cipher(algorithms.AES(self.key), modes.GCM(nonce)).encryptor()
        enc.authenticate_additional_data(aad)
        ct = enc.update(plaintext) + enc.finalize()
        tag = enc.tag[:TAG_SIZE]  # truncate 16-byte GCM tag to 12 bytes

        header = struct.pack("<II", INCOMING_MARKER, msg_length)
        self.sock.sendall(header + ct + tag)

    def recv_message(self) -> bytes:
        """Receive and decrypt one message from the server."""
        header_data = self._recv_exact(HEADER_SIZE)
        marker, msg_length = struct.unpack("<II", header_data)
        if marker != OUTGOING_MARKER:
            raise ValueError(f"bad outgoing marker: {marker:#010x}")
        if msg_length < HEADER_SIZE + TAG_SIZE:
            raise ValueError(f"message too short: {msg_length}")

        payload = self._recv_exact(msg_length - HEADER_SIZE)
        ct = payload[:-TAG_SIZE]
        tag = payload[-TAG_SIZE:]

        nonce = struct.pack("<Q", self.recv_nonce)
        self.recv_nonce += 1
        aad = struct.pack("<I", msg_length)

        dec = Cipher(
            algorithms.AES(self.key),
            modes.GCM(nonce, min_tag_length=TAG_SIZE),
        ).decryptor()
        dec.authenticate_additional_data(aad)
        plaintext = dec.update(ct)
        return plaintext + dec.finalize_with_tag(tag)

    def _recv_exact(self, n: int) -> bytes:
        data = bytearray()
        while len(data) < n:
            chunk = self.sock.recv(n - len(data))
            if not chunk:
                raise EOFError(
                    f"connection closed after {len(data)} of {n} expected bytes"
                )
            data.extend(chunk)
        return bytes(data)

    def recv_until_closed(self) -> bytes:
        """Read until the server closes the connection."""
        data = bytearray()
        self.sock.settimeout(2.0)
        try:
            while True:
                chunk = self.sock.recv(4096)
                if not chunk:
                    break
                data.extend(chunk)
        except (socket.timeout, ConnectionResetError):
            pass
        return bytes(data)


# ─── Tests ────────────────────────────────────────────────────────────────────

def test_encrypted_echo(
    test_binary: Path, scenario_catalog: dict[str, dict], tmp_path: Path
) -> None:
    """Basic encrypted echo using the default zero key."""
    require_supported(scenario_catalog, "encrypted_echo")

    def client_action(port: int) -> None:
        with connect_with_retry(port) as sock:
            client = EncryptedClient(sock)
            client.send_message(b"hello encrypted world")
            reply = client.recv_message()
            assert reply == b"hello encrypted world"

    payload = run_scenario(test_binary, "encrypted_echo", client_action, tmp_path)
    assert payload["numbers"]["connections"] == 1
    assert payload["numbers"]["closed_callbacks"] == 1


def test_encrypted_echo_with_key(
    test_binary: Path, scenario_catalog: dict[str, dict], tmp_path: Path
) -> None:
    """Encrypted echo using the non-zero test key."""
    require_supported(scenario_catalog, "encrypted_echo_with_key")

    def client_action(port: int) -> None:
        with connect_with_retry(port) as sock:
            client = EncryptedClient(sock, key=TEST_KEY)
            client.send_message(b"keyed message")
            reply = client.recv_message()
            assert reply == b"keyed message"

    payload = run_scenario(
        test_binary, "encrypted_echo_with_key", client_action, tmp_path
    )
    assert payload["numbers"]["connections"] == 1
    assert payload["numbers"]["closed_callbacks"] == 1


def test_encrypted_multiple_messages(
    test_binary: Path, scenario_catalog: dict[str, dict], tmp_path: Path
) -> None:
    """Send and receive multiple encrypted messages in sequence."""
    require_supported(scenario_catalog, "encrypted_multiple_messages")
    messages = [b"first", b"second", b"third"]

    def client_action(port: int) -> None:
        with connect_with_retry(port) as sock:
            client = EncryptedClient(sock)
            for msg in messages:
                client.send_message(msg)
                reply = client.recv_message()
                assert reply == msg

    payload = run_scenario(
        test_binary, "encrypted_multiple_messages", client_action, tmp_path
    )
    assert payload["numbers"]["connections"] == 1
    assert payload["numbers"]["closed_callbacks"] == 1


def test_encrypted_bad_marker(
    test_binary: Path, scenario_catalog: dict[str, dict], tmp_path: Path
) -> None:
    """Server must detect and reject a message with an invalid begin marker."""
    require_supported(scenario_catalog, "encrypted_bad_marker")

    def client_action(port: int) -> None:
        with connect_with_retry(port) as sock:
            # Send 30 bytes with a clearly wrong marker
            bad_header = struct.pack("<II", 0xDEADBEEF, 30)
            sock.sendall(bad_header + bytes(22))
            # Server will close the connection; drain any remaining bytes
            sock.settimeout(2.0)
            try:
                while sock.recv(1024):
                    pass
            except (socket.timeout, ConnectionResetError):
                pass

    payload = run_scenario(
        test_binary, "encrypted_bad_marker", client_action, tmp_path
    )
    assert payload["numbers"]["connections"] == 1
    assert payload["numbers"]["closed_callbacks"] == 1


def test_encrypted_bad_tag(
    test_binary: Path, scenario_catalog: dict[str, dict], tmp_path: Path
) -> None:
    """Server must detect and reject a message with a corrupted authentication tag."""
    require_supported(scenario_catalog, "encrypted_bad_tag")

    def client_action(port: int) -> None:
        with connect_with_retry(port) as sock:
            plaintext = b"tampered"
            # Build a valid message
            nonce = struct.pack("<Q", 0)
            msg_length = HEADER_SIZE + len(plaintext) + TAG_SIZE
            aad = struct.pack("<I", msg_length)
            enc = Cipher(algorithms.AES(ZERO_KEY), modes.GCM(nonce)).encryptor()
            enc.authenticate_additional_data(aad)
            ct = enc.update(plaintext) + enc.finalize()
            tag = enc.tag[:TAG_SIZE]
            # Corrupt the last byte of the tag
            bad_tag = tag[:-1] + bytes([tag[-1] ^ 0xFF])
            header = struct.pack("<II", INCOMING_MARKER, msg_length)
            sock.sendall(header + ct + bad_tag)
            # Server will close the connection
            sock.settimeout(2.0)
            try:
                while sock.recv(1024):
                    pass
            except (socket.timeout, ConnectionResetError):
                pass

    payload = run_scenario(
        test_binary, "encrypted_bad_tag", client_action, tmp_path
    )
    assert payload["numbers"]["connections"] == 1
    assert payload["numbers"]["closed_callbacks"] == 1

File: test/Server/test_EncryptedSocket.cpp
#include "test_EncryptedSocket.hpp"

#include <cstring>
#include <set>

#include "../../src/EncryptedSocket.hpp"

namespace {

// Hardcoded test key - Python client uses the exact same bytes.
// bytes(range(1, 17)) in Python.
constexpr uint8_t kTestKey[16] = {
	0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08,
	0x09, 0x0a, 0x0b, 0x0c, 0x0d, 0x0e, 0x0f, 0x10
};

constexpr int kMultiMessageCount = 3;

enum class EncryptedScenarioKind {
	EchoZeroKey,
	EchoWithKey,
	MultipleMessages,
	BadMarker,
	BadTag,
};

// ─── EncryptedScenarioListener ───────────────────────────────────────────────

class EncryptedScenarioListener : public BaseScenarioListener {
public:
	explicit EncryptedScenarioListener(EncryptedScenarioKind kind, const string &scenarioName)
		: BaseScenarioListener(scenarioName), kind(kind)
	{}

	void onConnected(const SP<Socket> &socket) override
	{
		const int id = idFor(socket);
		++connectedCount;
		logEvent("connected", {{"connection", std::to_string(id)}});

		auto enc = std::make_shared<EncryptedSocket>(socket);
		if (kind == EncryptedScenarioKind::EchoWithKey) {
			auto keyBuf = std::make_shared<bytes>(kTestKey, kTestKey + 16);
			enc->setKey(BytesView(keyBuf, 0, 16));
		}
		connections[id] = ConnState{std::move(enc), {}, 0};
	}

	void onData(const SP<Socket> &socket) override
	{
		++dataCallbackCount;
		const int id = idFor(socket);
		auto it = connections.find(id);
		if (it == connections.end()) return;

		ConnState &state = it->second;
		uint8_t *buf = nullptr;
		const size_t size = state.enc->read(buf);
		if (size == 0) return;  // incomplete message or error already queued

		const string msg(reinterpret_cast<const char *>(buf), size);
		state.received += msg;
		++state.msgCount;
		logEvent("data", {{"connection", std::to_string(id)}, {"msg", msg}});

		// Echo back
		auto msgBuf = std::make_shared<bytes>(buf, buf + size);
		state.enc->outputStart(size);
		state.enc->outputWrite(BytesView(msgBuf, 0, size));
		state.enc->outputFinish();
		echoedConnections.insert(id);
		logEvent("echoed", {{"connection", std::to_string(id)}, {"payload", msg}});

		checkCompletion();
	}

	ResultDocument buildDocument() const override
	{
		ResultDocument snap = BaseScenarioListener::buildDocument();
		for (const auto &e : connections)
			snap.stringLists["received_messages"].push_back(e.second.received);
		snap.numbers["echo_count"] = static_cast<long long>(echoedConnections.size());
		return snap;
	}

protected:
	void onSocketClosed(const SP<Socket> &, int) override
	{
		checkCompletion();
	}

	void onSocketError(const SP<Socket> &, int id, const string &error) override
	{
		if (kind == EncryptedScenarioKind::BadMarker || kind == EncryptedScenarioKind::BadTag) {
			logEvent("expected_error", {{"connection", std::to_string(id)}, {"error", error}});
			gotExpectedError = true;
			checkCompletion();
		} else {
			fail("unexpected socket close error: " + error);
		}
	}

private:
	struct ConnState {
		SP<EncryptedSocket> enc;
		string received;
		int msgCount = 0;
	};

	EncryptedScenarioKind kind;
	std::map<int, ConnState> connections;
	std::set<int> echoedConnections;
	bool gotExpectedError = false;

	void checkCompletion()
	{
		if (finished) return;
		switch (kind) {
		case EncryptedScenarioKind::EchoZeroKey:
		case EncryptedScenarioKind::EchoWithKey:
			if (connectedCount == 1 && echoedConnections.size() == 1 && closedCount == 1)
				pass("encrypted echo completed");
			break;

		case EncryptedScenarioKind::MultipleMessages: {
			auto it = connections.begin();
			if (connectedCount == 1 && closedCount == 1
				&& it != connections.end() && it->second.msgCount == kMultiMessageCount)
				pass("encrypted multiple messages completed");
			break;
		}

		case EncryptedScenarioKind::BadMarker:
		case EncryptedScenarioKind::BadTag:
			if (connectedCount == 1 && closedCount == 1 && gotExpectedError)
				pass("error correctly detected");
			break;
		}
	}
};

SP<ScenarioListenerBase> makeEncListener(EncryptedScenarioKind kind, const char *name)
{
	return SP<ScenarioListenerBase>(new EncryptedScenarioListener(kind, name));
}

} // namespace

const std::vector<ScenarioInfo> &encryptedSocketScenarios()
{
	static const std::vector<ScenarioInfo> catalog = {
		{
			"encrypted_echo",
			"Encrypt/decrypt one round-trip message with the default zero key",
			{"linux", "windows"},
			[] { return makeEncListener(EncryptedScenarioKind::EchoZeroKey, "encrypted_echo"); }
		},
		{
			"encrypted_echo_with_key",
			"Encrypt/decrypt one round-trip message with a non-zero key",
			{"linux", "windows"},
			[] { return makeEncListener(EncryptedScenarioKind::EchoWithKey, "encrypted_echo_with_key"); }
		},
		{
			"encrypted_multiple_messages",
			"Exchange multiple encrypted messages in sequence",
			{"linux", "windows"},
			[] { return makeEncListener(EncryptedScenarioKind::MultipleMessages, "encrypted_multiple_messages"); }
		},
		{
			"encrypted_bad_marker",
			"Server detects and rejects a message with an invalid begin marker",
			{"linux", "windows"},
			[] { return makeEncListener(EncryptedScenarioKind::BadMarker, "encrypted_bad_marker"); }
		},
		{
			"encrypted_bad_tag",
			"Server detects and rejects a message with a corrupted authentication tag",
			{"linux", "windows"},
			[] { return makeEncListener(EncryptedScenarioKind::BadTag, "encrypted_bad_tag"); }
		},
	};
	return catalog;
}
